### PR TITLE
Fashion MNIST tutorial

### PR DESCRIPTION
**Pull request by alsrgv**
Signed-off-by: Alex Sergeev <asergeev@uber.com>

### CODE REVIEW

File: fashion_mnist/INSTALL.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Creating VM for Fashion MNIST Tutorial
```
Comment (sblotner) [inline fashion_mnist/INSTALL.md]: VM for --> a VM for the

File: fashion_mnist/INSTALL.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
```
Comment (sblotner) [inline fashion_mnist/INSTALL.md:4]: These instructions describe how to install a VM for the [Fashion MNIST Tutorial](README.md).

File: fashion_mnist/INSTALL.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+## Prerequisites
+
+Installation assumes that VM will be based on *Ubuntu Server 16.04 LTS*.
```
Comment (sblotner) [inline fashion_mnist/INSTALL.md]: that VM --> that the VM

Comment (sblotner) [reply fashion_mnist/INSTALL.md]: Installation --> This installation

File: fashion_mnist/INSTALL.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+$ sudo bash
+# curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | apt-key add -
+# distribution=$(. /etc/os-release;echo $ID$VERSION_ID)
+# curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list > /etc/apt/sources.list.d/nvidia-docker.list
+# apt-get update
+# apt-get install -y nvidia-docker2
+# pkill -SIGHUP dockerd
+```
+
+## Build model docker
```
Comment (sblotner) [inline fashion_mnist/INSTALL.md]: Docker

File: fashion_mnist/INSTALL.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+```
+$ git clone https://github.com/horovod/tutorials && cd tutorials/fashion_mnist
+$ sudo docker build -t fashion_mnist_tutorial .
+```
+
+## Auto-start model docker
```
Comment (sblotner) [inline fashion_mnist/INSTALL.md]: Docker

File: fashion_mnist/README.md [hunk 6]
```diff
@@ -0,0 +1,394 @@
+# Fashion MNIST Tutorial
+
+In this tutorial, you will learn how to apply Horovod to a [WideResNet](https://arxiv.org/abs/1605.07146) model, trained on Fashion MNIST dataset.
+
+If this is an in-person session, hosts will set up VM for you and provide you credentials to Jupyter Lab.  If you're working on this tutorial on your own, please follow installation instructions in [INSTALL.md](INSTALL.md).
```
Comment (sblotner) [inline fashion_mnist/README.md:7]: Recommend putting this paragraph under a ## Prerequisites heading.

File: fashion_mnist/README.md [hunk 7]
```diff
@@ -0,0 +1,394 @@
+# Fashion MNIST Tutorial
+
+In this tutorial, you will learn how to apply Horovod to a [WideResNet](https://arxiv.org/abs/1605.07146) model, trained on Fashion MNIST dataset.
```
Comment (sblotner) [inline fashion_mnist/README.md]: trained on --> trained on the

File: fashion_mnist/README.md [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+
+If this is an in-person session, hosts will set up VM for you and provide you credentials to Jupyter Lab.  If you're working on this tutorial on your own, please follow installation instructions in [INSTALL.md](INSTALL.md).
+
+Let's begin!
+
+## Connect to Jupyter Lab
```
Comment (sblotner) [inline fashion_mnist/README.md:11]: How do you connect to Jupyter lab? If it's run or click X, recommend specifying that here.

File: fashion_mnist/README.md [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+
+Once you open Jupyter Lab, you will see screen similar to this:
```
Comment (sblotner) [inline fashion_mnist/README.md]: Once --> When
see screen --> see a screen

File: fashion_mnist/README.md [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+
+![image](https://user-images.githubusercontent.com/16640218/53517400-5d2f0480-3a83-11e9-8db9-268037fc6c2f.png)
+
+In this lab, we will use Terminal and File Editor features.
```
Comment (sblotner) [inline fashion_mnist/README.md]: use --> use the

File: fashion_mnist/README.md [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+
+## Explore model files
+
+On the left hand side, you will see two Python files: `fashion_mnist.py` and `fashion_mnist_solution.py`.
+
+<img src="https://user-images.githubusercontent.com/16640218/53517505-9f584600-3a83-11e9-8e1f-9099e9690299.png" width="300"></img>
+
+The first file contains the Keras model that does not have any Horovod code, while the second one has all the Horovod features added.  In this tutorial, we will guide you to transform `fashion_mnist.py` into `fashion_mnist_solution.py` step-by-step.
+
+Why Keras?  We chose Keras due to it's simplicity, and the fact that it will be the way to define models in TensorFlow 2.0.
```
Comment (sblotner) [inline fashion_mnist/README.md]: it's --> its

File: fashion_mnist/README.md [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+
+## Run fashion_mnist.py
+
+Before we go into modifications requires to scale our WideResNet model, let's run a single-GPU version of the model.
```
Comment (sblotner) [inline fashion_mnist/README.md]: requires --> required

File: fashion_mnist/README.md [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+
+In the Launcher, click Terminal button:
```
Comment (sblotner) [inline fashion_mnist/README.md]: click --> click the

File: fashion_mnist/README.md [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+config.gpu_options.allow_growth = True
+config.gpu_options.visible_device_list = str(hvd.local_rank())
+K.set_session(tf.Session(config=config))
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53518149-4689ad00-3a85-11e9-9f59-f22eeba05e73.png)
+
+### Broadcast starting epoch from first worker to everyone else
+
+In `fashion_mnist.py`, we're using filename of the last checkpoint to determine epoch to resume training from in case of a failure:
```
Comment (sblotner) [inline fashion_mnist/README.md]: using --> using the 
determine --> determine the

File: fashion_mnist/README.md [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+# Horovod: pin GPU to be used to process local rank (one GPU per process)
+config = tf.ConfigProto()
+config.gpu_options.allow_growth = True
+config.gpu_options.visible_device_list = str(hvd.local_rank())
+K.set_session(tf.Session(config=config))
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53518149-4689ad00-3a85-11e9-9f59-f22eeba05e73.png)
+
+### Broadcast starting epoch from first worker to everyone else
```
Comment (sblotner) [inline fashion_mnist/README.md]: starting --> the starting
from first --> from the first

File: fashion_mnist/README.md [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+```python
+# If set > 0, will resume training from a given checkpoint.
+resume_from_epoch = 0
+for try_epoch in range(args.epochs, 0, -1):
+    if os.path.exists(args.checkpoint_format.format(epoch=try_epoch)):
+        resume_from_epoch = try_epoch
+        break
+```
+
+As you scale your workload to multi-node, some of your workers may not have access to the filesystem containing the checkpoint.  For that reason, we make the first worker to determine epoch to restart from, and *broadcast* that information to the rest of the workers.
```
Comment (sblotner) [inline fashion_mnist/README.md]: determine epoch --> determine the epoch

File: fashion_mnist/README.md [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+
+This requires adding the following code:
```
Comment (sblotner) [inline fashion_mnist/README.md]: To broadcast the starting epoch from the first worker, add the following code:

File: fashion_mnist/README.md [hunk 18]
```diff
@@ -0,0 +1,394 @@
...
+
+```python
+# Horovod: broadcast resume_from_epoch from rank 0 (which will have
+# checkpoints) to other ranks.
+resume_from_epoch = hvd.broadcast(resume_from_epoch, 0, name='resume_from_epoch')
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53534072-2de3bc00-3ab2-11e9-8cf1-7531542e3202.png)
+
+### Print verbose logs only on first worker
```
Comment (sblotner) [inline fashion_mnist/README.md]: on --> on the

File: fashion_mnist/README.md [hunk 19]
```diff
@@ -0,0 +1,394 @@
...
+
+Horovod uses MPI to run model training workers.  By default, MPI aggregates output from all workers.  In order to reduce clutter, it's recommended to write logs only on first worker.
```
Comment (sblotner) [inline fashion_mnist/README.md]: In order to --> To 
it's recommended to write --> we recommend that you write 
on first --> on the first

File: fashion_mnist/README.md [hunk 20]
```diff
@@ -0,0 +1,394 @@
...
+Replace `verbose = 1` with the following code:
+
+```python
+# Horovod: print logs on the first worker.
+verbose = 1 if hvd.rank() == 0 else 0
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53534314-2244c500-3ab3-11e9-95ef-e7e7b282ab4f.png)
+
+### Read checkpoint only on first worker
```
Comment (sblotner) [inline fashion_mnist/README.md]: on --> on the

File: fashion_mnist/README.md [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+
+For the same reason as above, we read checkpoint only the first worker and *broadcast* initial state to other workers.
```
Comment (sblotner) [inline fashion_mnist/README.md]: read checkpoint only --> read the checkpoint only with
initial --> the initial

File: fashion_mnist/README.md [hunk 22]
```diff
@@ -0,0 +1,394 @@
...
+    model = hvd.load_model(args.checkpoint_format.format(epoch=resume_from_epoch))
+else:
+    ...
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53534410-9717ff00-3ab3-11e9-86eb-1bf8299416d2.png)
+
+### Adjust learning rate and add Distributed Optimizer
+
+Horovod uses an operation that averages gradients across workers.  Gradient averaging typically requires corresponding increase in learning rate to make bigger steps in the direction of a higher-quality gradient.
```
Comment (sblotner) [inline fashion_mnist/README.md]: requires --> requires a

File: fashion_mnist/README.md [hunk 23]
```diff
@@ -0,0 +1,394 @@
...
+
+# Horovod: add Horovod Distributed Optimizer.
+opt = hvd.DistributedOptimizer(opt)
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53534579-52409800-3ab4-11e9-971e-f7def73c7b36.png)
+
+### Add BroadcastGlobalVariablesCallback
+
+In previous section, we mentioned that first worker would broadcast parameters to the rest of the workers.  We will use `horovod.keras.BroadcastGlobalVariablesCallback` to make this happen.
```
Comment (sblotner) [inline fashion_mnist/README.md]: previous --> the previous 
first --> the first

File: fashion_mnist/README.md [hunk 24]
```diff
@@ -0,0 +1,394 @@
...
+
+Add `BroadcastGlobalVariablesCallback` as the first element of `callbacks` list:
```
Comment (sblotner) [inline fashion_mnist/README.md]: of --> of the

File: fashion_mnist/README.md [hunk 25]
```diff
@@ -0,0 +1,394 @@
...
+    hvd.callbacks.BroadcastGlobalVariablesCallback(0),
+
+    ...
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53535043-1dcddb80-3ab6-11e9-9911-1eb33a1f531c.png)
+
+### Add learning rate warmup
+
+Many models are sensitive to using large learning rate immediately after initialization and can benefit from learning rate warmup.  The idea is to start training with lower LR and gradually raise it to target LR over few epochs.  Horovod has convenient `LearningRateWarmupCallback` for Keras API that implements that logic.
```
Comment (sblotner) [inline fashion_mnist/README.md]: large learning rate --> a large learning rate (LR)
raise it to --> raise it to a 
over few --> over a few
Horovod has --> Horovod has the 
for Keras API --> for the Keras API

File: fashion_mnist/README.md [hunk 26]
```diff
@@ -0,0 +1,394 @@
...
+
+Since we're already using `LearningRateScheduler` in this code, and it modifies learning rate along with `LearningRateWarmupCallback`, there is a possibility of a conflict.  In order to avoid such conflict, we will swap out `LearningRateScheduler` with Horovod `LearningRateScheduleCallback`.
+
+We will replace the following code:
```
Comment (sblotner) [inline fashion_mnist/README.md]: We will replace --> Replace

File: fashion_mnist/README.md [hunk 27]
```diff
@@ -0,0 +1,394 @@
...
+    hvd.callbacks.LearningRateScheduleCallback(start_epoch=15, end_epoch=25, multiplier=1e-1),
+    hvd.callbacks.LearningRateScheduleCallback(start_epoch=25, end_epoch=35, multiplier=1e-2),
+    hvd.callbacks.LearningRateScheduleCallback(start_epoch=35, multiplier=1e-3),
+
+    ...
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53535420-98e3c180-3ab7-11e9-8780-9258081f66c5.png)
+
+Since we've added new `args.warmup_epochs` argument, we should register it:
```
Comment (sblotner) [inline fashion_mnist/README.md]: added --> added a

File: fashion_mnist/README.md [hunk 28]
```diff
@@ -0,0 +1,394 @@
...
+if hvd.rank() == 0:
+    callbacks.append(keras.callbacks.ModelCheckpoint(args.checkpoint_format))
+    callbacks.append(keras.callbacks.TensorBoard(args.log_dir))
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53535616-4e167980-3ab8-11e9-82d9-82e431dfd621.png)
+
+### Modify training loop to execute fewer steps per epoch
+
+In order to speed up training, we will execute fewer steps of distributed training.  In order to keep the total number of examples processed during the training the same, we will do `num_steps / N` steps, where `num_steps` is the original number of steps, and `N` is the total number of workers.
```
Comment (sblotner) [inline fashion_mnist/README.md]: In order to --> To 
(^^ for both sentences)

File: fashion_mnist/README.md [hunk 29]
```diff
@@ -0,0 +1,394 @@
...
+
+We will also speed-up validation by validating `3 * num_validation_steps / N` steps on each worker.  Multiplier **3** provides over-sampling of validation data helps to increase probability that every validation example will be evaluated.
```
Comment (sblotner) [inline fashion_mnist/README.md]: speed-up --> speed up
Multiplier --> The multiplier 
data helps --> data, which helps
increase --> increase the

File: fashion_mnist/README.md [hunk 30]
```diff
@@ -0,0 +1,394 @@
...
+                    validation_steps=3 * len(test_iter) // hvd.size())
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53536410-283ea400-3abb-11e9-8742-05921b0795de.png)
+
+### Average validation results among workers
+
+Since we're not validating full dataset on each worker anymore, each worker will have different validation results.  To improve validation metric quality and reduce variance, we will average validation results among all workers.
+
+To do so, we will inject `MetricAverageCallback` after `BroadcastGlobalVariablesCallback`:
```
Comment (sblotner) [inline fashion_mnist/README.md]: remove "we will"

File: fashion_mnist/README.md [hunk 31]
```diff
@@ -0,0 +1,394 @@
...
+    hvd.callbacks.MetricAverageCallback(),
+    
+    ...
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53536553-b2870800-3abb-11e9-88f9-1a2758bd25dd.png)
+
+### Check your work
+
+Congratulations!  If you made it this far, your `fashion_mnist.py` should now be fully distributed.  In order to verify, you can run the following command in the terminal, which should produce no output:
```
Comment (sblotner) [inline fashion_mnist/README.md]: In order to --> To

File: fashion_mnist/README.md [hunk 32]
```diff
@@ -0,0 +1,394 @@
...
+
+Now, run distributed `fashion_mnist.py` using:
+
+```
+$ horovodrun -np 4 python fashion_mnist.py --log-dir distributed
+```
+
+![image](https://user-images.githubusercontent.com/16640218/53536888-da2aa000-3abc-11e9-9083-43060634433c.png)
+
+After a few minutes, you should see training progress.  It will be faster compared to single-GPU model:
```
Comment (sblotner) [inline fashion_mnist/README.md]: compared to --> compared to the

File: fashion_mnist/README.md [hunk 33]
```diff
@@ -0,0 +1,394 @@
...
+
+![image](https://user-images.githubusercontent.com/16640218/53536956-270e7680-3abd-11e9-8f3b-acbe9bbfd085.png)
+
+## Monitor training progress
+
+In order to monitor training progress and compare runs, we will use TensorBoard.
```
Comment (sblotner) [inline fashion_mnist/README.md]: In order to --> To

File: fashion_mnist/README.md (hunk 7 above)
Comment (sblotner) [reply fashion_mnist/README.md]: Consider linking to https://github.com/zalandoresearch/fashion-mnist

**Review by sblotner — COMMENTED** [review]
This is great! Really like how you explain why you're making each change along the way.

**Comment by alsrgv** [comment]
@sblotner, thanks for all the copyedits, it's amazing!

**Review by sblotner — APPROVED** [review]
